packages/block-editor/src/components/block-draggable/index.js

```diff
@@ -61,6 +86,64 @@ const BlockDraggable = ( {
 		};
 	}, [] );
 
+	// Find the root of the editor iframe.
+	const blockRef = useBlockRef( clientIds[ 0 ] );
+	const editorRoot = blockRef.current?.closest( 'body' );
+
+	// Add a dragover event listener to the editor root to track the blocks being dragged over.
+	// The listener has to be inside the editor iframe otherwise the target isn't accessible.
+	// Check if the dragged blocks are allowed inside the target. If not, grey out the draggable.
+	useEffect( () => {
+		if ( ! editorRoot ) {
+			return;
+		}
+
+		const onDragOver = ( event ) => {
```

Comment:
I think this callback might need to be throttled similar to the code in `useBlockDropZone` as this currently it fires very frequently:

https://github.com/WordPress/gutenberg/assets/14988353/0ece29d6-b06e-4555-8ad7-c0ecd1ca76ce

Reply:
Hrm yeah I'm wondering how best to do that given the inability to use hooks inside the effect callback 🤔

Reply:
I think I ran into this issue when testing, it seemed to crash in Safari and I couldn't move the chip any more.

Reply:
Oo wait I see there's a `throttle` util in compose that we may be able to leverage, will try that!

Reply:
Ok the throttle seems to make it better! I set it to the same as [draggable](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/draggable/index.tsx#L208).

Reply:
That's a little better, but it's still lagging a bit for me with the throttle at `16` (it's more visible when using CPU throttling in dev tools). I think a couple of things might be at play:

* Since the `useEffect` doesn't use a dependency array, it seems that it's being called quite frequently, and potentially adding (and removing) event listeners too frequently? For example, if I increase the throttle time to `2000`, it still appears to be firing rapidly
* In `draggable`, `16ms` is needed because it's updating the position of the drag chip. Here, I'm wondering if we can use a bigger delay / throttle, if that could smooth over the flicker when dragging between image blocks in a gallery block? I.e. if we had it at `200ms` to match `useBlockDropZone` then if folks are dragging quickly between blocks, hopefully it shouldn't flicker as much?

Would it work to maybe move defining the `onDragOver` callback to be outside of the `useEffect` (possibly in a `useCallback`) so that the `useEffect` could have a dependency array and only attach the event listener once? Apologies if you've already tried that!

Reply:
So I had a think about this 😄 and realised that we actually don't need to use local state for `targetClientId`, and we can use a dependency array for the effect, which improves things a little, but it still has to re-render whenever either `draggedBlockNames` or `visibleInserter` change. Moving the callback out of the effect makes no difference, because it calls whichever version of the callback was attached to the event listener when the effect ran, so if we want the values inside the callback to change the effect has to re-run.

I also tried increasing the throttle time and it doesn't seem to make things any worse, so might leave it at 200 or so. I just pushed an update, let me know if it improves things at all!

Reply:
Nice work! That appears to have fixed up the performance issues for me. Subjectively, without CPU throttling, it's feeling smooth now, and with CPU throttling, it seems much the same as `trunk` to me now 🎉